File: test_safety_and_reasoning.py
"""
Test script to demonstrate Safety Filtering and Reasoning Controls
in both Python and JavaScript implementations.
"""

from PolliLibPy.text_to_text import TextToText
from PolliLibPy.text_to_image import TextToImage


def test_reasoning_controls():
    """Test reasoning_effort parameter with different levels"""
    print("=" * 70)
    print("TESTING REASONING CONTROLS")
    print("=" * 70)

    generator = TextToText()

    # Test prompt that benefits from deep reasoning
    prompt = "Plan a 3-day trip to Paris with a budget of $1500"

    reasoning_levels = ["minimal", "low", "medium", "high"]

    for level in reasoning_levels:
        print(f"\n{'='*70}")
        print(f"Testing reasoning_effort: {level}")
        print(f"{'='*70}")

        result = generator.chat(
            messages=[{"role": "user", "content": prompt}],
            model="openai",
            reasoning_effort=level,
            max_tokens=200,
            temperature=0.7
        )

        if result['success']:
            print(f"\n✓ Response with {level} reasoning:")
            print(f"  {result['response'][:300]}...")
            print(f"\n  Usage: {result.get('usage', {})}")
        else:
            print(f"✗ Error: {result['error']}")

    print("\n" + "=" * 70)
    print("Reasoning controls test complete!")
    print("=" * 70)


def test_safety_filtering_text():
    """Test safe parameter for text generation"""
    print("\n\n" + "=" * 70)
    print("TESTING SAFETY FILTERING - TEXT GENERATION")
    print("=" * 70)

    generator = TextToText()

    # Test with safe mode enabled
    print("\n1. Testing with safe=True:")
    print("-" * 70)

    result = generator.chat(
        messages=[{"role": "user", "content": "Tell me a family-friendly joke"}],
        model="openai",
        safe=True,
        temperature=0.7
    )

    if result['success']:
        print(f"✓ Response with safety filter enabled:")
        print(f"  {result['response']}")
        print(f"  Safety Check: {result.get('safetyCheck', {})}")
    else:
        print(f"✗ Error: {result['error']}")

    # Test with safe mode disabled
    print("\n2. Testing with safe=False (default):")
    print("-" * 70)

    result = generator.chat(
        messages=[{"role": "user", "content": "Tell me a joke about programming"}],
        model="openai",
        safe=False,
        temperature=0.7
    )

    if result['success']:
        print(f"✓ Response without safety filter:")
        print(f"  {result['response']}")
    else:
        print(f"✗ Error: {result['error']}")

    print("\n" + "=" * 70)
    print("Text safety filtering test complete!")
    print("=" * 70)


def test_safety_filtering_image():
    """Test safe parameter for image generation"""
    print("\n\n" + "=" * 70)
    print("TESTING SAFETY FILTERING - IMAGE GENERATION")
    print("=" * 70)

    generator = TextToImage()

    # Test with safe mode enabled
    test_prompts = [
        "a beautiful sunset over mountains",
        "a family having a picnic in the park",
        "cute puppies playing with toys"
    ]

    print("\nTesting safe mode for image generation:")
    print("-" * 70)

    for prompt in test_prompts:
        print(f"\nPrompt: '{prompt}'")

        result = generator.generate_image(
            prompt=prompt,
            safe=True,
            model="turbo",
            width=512,
            height=512
        )

        if result['success']:
            print(f"  ✓ PASSED safety filter")
            print(f"  Inference time: {result['inference_time']:.2f}s")
        else:
            print(f"  ✗ BLOCKED: {result.get('message', result.get('error'))}")

    print("\n" + "=" * 70)
    print("Image safety filtering test complete!")
    print("=" * 70)


def test_combined_features():
    """Test using both safety and reasoning controls together"""
    print("\n\n" + "=" * 70)
    print("TESTING COMBINED FEATURES (Safety + Reasoning)")
    print("=" * 70)

    generator = TextToText()

    print("\nGenerating a detailed, family-friendly travel guide:")
    print("-" * 70)

    result = generator.chat(
        messages=[{
            "role": "user",
            "content": "Create a detailed family-friendly itinerary for a day in Disney World"
        }],
        model="openai",
        reasoning_effort="high",  # Use deep reasoning for detailed planning
        safe=True,  # Ensure family-friendly content
        temperature=0.7,
        max_tokens=500
    )

    if result['success']:
        print(f"✓ Generated itinerary:")
        print(f"\n{result['response']}")
        print(f"\nUsage: {result.get('usage', {})}")
        print(f"Safety Check: {result.get('safetyCheck', {})}")
    else:
        print(f"✗ Error: {result['error']}")

    print("\n" + "=" * 70)
    print("Combined features test complete!")
    print("=" * 70)


if __name__ == "__main__":
    print("\n" + "=" * 70)
    print("SAFETY FILTERING & REASONING CONTROLS TEST SUITE")
    print("Testing both Python implementations")
    print("=" * 70)

    # Run all tests
    try:
        test_reasoning_controls()
        test_safety_filtering_text()
        test_safety_filtering_image()
        test_combined_features()

        print("\n\n" + "=" * 70)
        print("ALL TESTS COMPLETED SUCCESSFULLY!")
        print("=" * 70)
        print("\nSummary:")
        print("  ✓ Reasoning Controls (minimal, low, medium, high) - Working")
        print("  ✓ Safety Filtering for Text Generation - Working")
        print("  ✓ Safety Filtering for Image Generation - Working")
        print("  ✓ Combined Features - Working")
        print("\nBoth Python and JavaScript implementations support:")
        print("  - reasoning_effort parameter for text generation")
        print("  - safe parameter for text and image generation")
        print("=" * 70)

    except Exception as e:
        print(f"\n✗ Test suite failed with error: {e}")
        import traceback
        traceback.print_exc()

File: PolliLibJS/test_safety_and_reasoning.js
/**
 * Test script to demonstrate Safety Filtering and Reasoning Controls
 * in JavaScript implementation.
 */

const { TextToText } = require('./text-to-text');
const { TextToImage } = require('./text-to-image');

/**
 * Test reasoning_effort parameter with different levels
 */
async function testReasoningControls() {
    console.log("=".repeat(70));
    console.log("TESTING REASONING CONTROLS");
    console.log("=".repeat(70));

    const generator = new TextToText();

    // Test prompt that benefits from deep reasoning
    const prompt = "Plan a 3-day trip to Paris with a budget of $1500";

    const reasoningLevels = ["minimal", "low", "medium", "high"];

    for (const level of reasoningLevels) {
        console.log("\n" + "=".repeat(70));
        console.log(`Testing reasoning_effort: ${level}`);
        console.log("=".repeat(70));

        const result = await generator.chat({
            messages: [{ role: "user", content: prompt }],
            model: "openai",
            reasoningEffort: level,
            maxTokens: 200,
            temperature: 0.7
        });

        if (result.success) {
            console.log(`\n✓ Response with ${level} reasoning:`);
            console.log(`  ${result.response.substring(0, 300)}...`);
            console.log(`\n  Usage:`, result.usage || {});
        } else {
            console.log(`✗ Error: ${result.error}`);
        }
    }

    console.log("\n" + "=".repeat(70));
    console.log("Reasoning controls test complete!");
    console.log("=".repeat(70));
}

/**
 * Test safe parameter for text generation
 */
async function testSafetyFilteringText() {
    console.log("\n\n" + "=".repeat(70));
    console.log("TESTING SAFETY FILTERING - TEXT GENERATION");
    console.log("=".repeat(70));

    const generator = new TextToText();

    // Test with safe mode enabled
    console.log("\n1. Testing with safe=true:");
    console.log("-".repeat(70));

    let result = await generator.chat({
        messages: [{ role: "user", content: "Tell me a family-friendly joke" }],
        model: "openai",
        safe: true,
        temperature: 0.7
    });

    if (result.success) {
        console.log(`✓ Response with safety filter enabled:`);
        console.log(`  ${result.response}`);
        console.log(`  Safety Check:`, result.safetyCheck || {});
    } else {
        console.log(`✗ Error: ${result.error}`);
    }

    // Test with safe mode disabled
    console.log("\n2. Testing with safe=false (default):");
    console.log("-".repeat(70));

    result = await generator.chat({
        messages: [{ role: "user", content: "Tell me a joke about programming" }],
        model: "openai",
        safe: false,
        temperature: 0.7
    });

    if (result.success) {
        console.log(`✓ Response without safety filter:`);
        console.log(`  ${result.response}`);
    } else {
        console.log(`✗ Error: ${result.error}`);
    }

    console.log("\n" + "=".repeat(70));
    console.log("Text safety filtering test complete!");
    console.log("=".repeat(70));
}

/**
 * Test safe parameter for image generation
 */
async function testSafetyFilteringImage() {
    console.log("\n\n" + "=".repeat(70));
    console.log("TESTING SAFETY FILTERING - IMAGE GENERATION");
    console.log("=".repeat(70));

    const generator = new TextToImage();

    // Test with safe mode enabled
    const testPrompts = [
        "a beautiful sunset over mountains",
        "a family having a picnic in the park",
        "cute puppies playing with toys"
    ];

    console.log("\nTesting safe mode for image generation:");
    console.log("-".repeat(70));

    for (const prompt of testPrompts) {
        console.log(`\nPrompt: '${prompt}'`);

        const result = await generator.generateImage({
            prompt,
            safe: true,
            model: "turbo",
            width: 512,
            height: 512
        });

        if (result.success) {
            console.log(`  ✓ PASSED safety filter`);
            console.log(`  Inference time: ${result.inferenceTime.toFixed(2)}s`);
        } else {
            console.log(`  ✗ BLOCKED: ${result.message || result.error}`);
        }
    }

    console.log("\n" + "=".repeat(70));
    console.log("Image safety filtering test complete!");
    console.log("=".repeat(70));
}

/**
 * Test using both safety and reasoning controls together
 */
async function testCombinedFeatures() {
    console.log("\n\n" + "=".repeat(70));
    console.log("TESTING COMBINED FEATURES (Safety + Reasoning)");
    console.log("=".repeat(70));

    const generator = new TextToText();

    console.log("\nGenerating a detailed, family-friendly travel guide:");
    console.log("-".repeat(70));

    const result = await generator.chat({
        messages: [{
            role: "user",
            content: "Create a detailed family-friendly itinerary for a day in Disney World"
        }],
        model: "openai",
        reasoningEffort: "high",  // Use deep reasoning for detailed planning
        safe: true,  // Ensure family-friendly content
        temperature: 0.7,
        maxTokens: 500
    });

    if (result.success) {
        console.log(`✓ Generated itinerary:`);
        console.log(`\n${result.response}`);
        console.log(`\nUsage:`, result.usage || {});
        console.log(`Safety Check:`, result.safetyCheck || {});
    } else {
        console.log(`✗ Error: ${result.error}`);
    }

    console.log("\n" + "=".repeat(70));
    console.log("Combined features test complete!");
    console.log("=".repeat(70));
}

/**
 * Main test runner
 */
async function main() {
    console.log("\n" + "=".repeat(70));
    console.log("SAFETY FILTERING & REASONING CONTROLS TEST SUITE");
    console.log("Testing JavaScript implementation");
    console.log("=".repeat(70));

    try {
        await testReasoningControls();
        await testSafetyFilteringText();
        await testSafetyFilteringImage();
        await testCombinedFeatures();

        console.log("\n\n" + "=".repeat(70));
        console.log("ALL TESTS COMPLETED SUCCESSFULLY!");
        console.log("=".repeat(70));
        console.log("\nSummary:");
        console.log("  ✓ Reasoning Controls (minimal, low, medium, high) - Working");
        console.log("  ✓ Safety Filtering for Text Generation - Working");
        console.log("  ✓ Safety Filtering for Image Generation - Working");
        console.log("  ✓ Combined Features - Working");
        console.log("\nJavaScript implementation supports:");
        console.log("  - reasoningEffort parameter for text generation");
        console.log("  - safe parameter for text and image generation");
        console.log("=".repeat(70));

    } catch (error) {
        console.log(`\n✗ Test suite failed with error: ${error.message}`);
        console.error(error);
    }
}

// Run tests if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = {
    testReasoningControls,
    testSafetyFilteringText,
    testSafetyFilteringImage,
    testCombinedFeatures
};
